restructured networksimplex.py and added test_networksimplex.py

harshal-dupare opened the pull request:
Restructured `networksimplex.py` and added a class to handle the internal function and essential data, and for testing added `test_networksimplex.py` with few new test cases taken from [Google OR-Tools](https://developers.google.com/optimization/flow/mincostflow) ( do check out google colab notebook option to run on an example )

Closes #4660

dschult commented:
Wow!  This is Great!!    Thanks!!!

I haven't had a chance to go through it yet, but I think (if you want) you could remove the leading underscore from the method names since the class has an underscore to indicate private.  But that's up to you.
I look forward to reading it more carefully. :}

harshal-dupare commented:
Sure, will remove the extra underscore : D. I had one question, I left the function `add_entry` on line `622` (in this PR) ( `555` in main ) be as it is....why? as it is just adding the flow dictionary entry and not updating any of the other essential data. Is that ok?

rossbar commented on networkx/algorithms/flow/tests/test_networksimplex.py:
```diff
@@ -0,0 +1,401 @@
+                flowCost += flowDict[u][v] * G[u][v]["weight"]
+        return flowCost
+
+    def test_infinite_demand_raise(self):
+        G = nx.DiGraph()
+        inf = float("inf")
+        G.add_node("a", demand=inf)
+        G.add_node("b", demand=-5)
+        G.add_node("c", demand=50000000)
+        G.add_node("d", demand=-49999995)
+        G.add_edge("a", "b", weight=3, capacity=4)
+        G.add_edge("a", "c", weight=6, capacity=10)
+        G.add_edge("b", "d", weight=1, capacity=9)
+        G.add_edge("c", "d", weight=2, capacity=5)
+        pytest.raises(nx.NetworkXError, nx.network_simplex, G)
```
It seems like this same basic example graph (with minor modifications) is used for the next 6 or so tests. The redundancy could really be cut down by wrapping the basic graph generation up in a fixture, and passing that into each of the tests; e.g. something like:

```suggestion
    @pytest.fixture
    def simple_flow_graph():
        G = nx.DiGraph()
        G.add_node("a", demand=5)
        G.add_node("b", demand=-5)
        G.add_node("c", demand=50000000)
        G.add_node("d", demand=-49999995)
        G.add_edge("a", "b", weight=3, capacity=4)
        G.add_edge("a", "c", weight=6, capacity=10)
        G.add_edge("b", "d", weight=1, capacity=9)
        G.add_edge("c", "d", weight=2, capacity=5)
        return G

    def test_infinite_demand_raise(self, simple_flow_graph):
        G = simple_flow_graph
        inf = float("inf")
        nx.set_node_attributes(G, {"a": {"demand": inf}})
        pytest.raises(nx.NetworkXError, nx.network_simplex, G)

```

Then the next few tests can be rewritten to use the fixture in a similar manner.

harshal-dupare commented on networkx/algorithms/flow/tests/test_networksimplex.py:
```diff
@@ -0,0 +1,401 @@
+import os
+
+
+class TestNetworkSimplex:
```
Thanks for the review @rossbar. I made both the suggested changes in `test_networksimplex.py` and it looks much better now!

rossbar commented on networkx/algorithms/flow/networksimplex.py:
```diff
@@ -10,6 +10,321 @@
+            False  # False untill initilize_spanning_tree() is called
+        )
+
+    def initilize_spanning_tree(self, n, faux_inf):
+        self.edge_count = len(self.edge_indices)  # number of edges
+        self.edge_flow = list(
+            chain(repeat(0, self.edge_count), (abs(d) for d in self.node_demands))
+        )  # edge flows
+        self.node_potentials = [
+            faux_inf if d <= 0 else -faux_inf for d in self.node_demands
+        ]  # node potentials
+        self.parent = list(chain(repeat(-1, n), [None]))  # parent nodes
+        self.parent_edge = list(
+            range(self.edge_count, self.edge_count + n)
+        )  # edges to parents
+        self.subtree_size = list(chain(repeat(1, n), [n + 1]))  # subtree sizes
+        self.next_node_dft = list(
+            chain(range(1, n), [-1, 0])
+        )  # next nodes in depth-first thread
+        self.prev_node_dft = list(range(-1, n))  # previous nodes in depth-first thread
+        self.last_descendent_dft = list(
+            chain(range(n), [n - 1])
+        )  # last descendants in depth-first thread
+        self._spanning_tree_initilized = True  # True only if all the assignments pass
```
Same comment here about the comments: most are unnecessary with the more descriptive attribute names.

rossbar commented on networkx/algorithms/flow/networksimplex.py:
```diff
@@ -180,43 +495,23 @@ def network_simplex(G, demand="demand", capacity="capacity", weight="weight"):
 
-    for e, c in zip(E, C):
-        if abs(c) == inf:
+    ###########################################################################
+    # Quick Error Detection
+    ###########################################################################
```
Could the error/infeasibility detection + this initial setup be moved inside the `_DataEssentialsAndFunctions` constructor? Was there a specific reason why the error checking was left outside of the class? If not, and it's possible to move all of this checking/initialization inside the constructor/setup methods, then it might be possible to simplify the `_DataEssentialsAndFunctions` signature. For example, it'd be nice if `multigraph = G.is_multigraph()` were only necessary inside the class, eliminating the need to pass around the `multigraph` parameter.

harshal-dupare replied:
I thought of `_DataEssentialsAndFunctions` as a way of abstracting away the data and related functions. Hence, I did not consider the possibility of including checks within it. But we can certainly include initial checks in the `_DataEssentialsAndFunctions` constructor. But then we would have to choose between letting the infeasibility and unboundedness checks ( they need to be after pivot loop) be in `network_simplex` and other checks in `_DataEssentialsAndFunctions`, which seems to be a very scattered way of managing checks (or is it a better way? if so, we can include it there), or we can have methods in `_DataEssentialsAndFunctions` for all of these checks (or is there any other better way?). And about the  Initialization part, it would be best to include it in the `initialize_spanning_tree` call which would make that part of `network_simplex` clean. the problem with `multigraph` is that it is also required in flow cost calculation but we can still get rid of the multigraph parameter by `self.multigraph = G.is_multigraph()` in `_DataEssentialsAndFunctions` constructor and using `DEAF.multigraph` instead of `multigraph` in `network_simplex`.

rossbar replied:
> But then we would have to choose between letting the infeasibility and unboundedness checks ( they need to be after pivot loop) be in network_simplex and other checks in _DataEssentialsAndFunctions, which seems to be a very scattered way of managing checks (or is it a better way? if so, we can include it there), or we can have methods in _DataEssentialsAndFunctions for all of these checks (or is there any other better way?)

Thanks for the explaining the thought process here - I was mostly curious if there were a *simple* way of reducing the number of parameters in the constructor, but it's a minor point and not worth spending cycles on IMO. I think the organization you've chosen is entirely fine and would vote to leave it as is :)

> And about the Initialization part, it would be best to include it in the initialize_spanning_tree call which would make that part of network_simplex clean.

I think I agree - it'd be great to get rid of the pre-initialization-initializations (i.e. defining the class attributes and setting them to `None`) and I definitely would like to get rid of the `_spanning_tree_initialized` flag, which seems like the way this is headed.

> the problem with multigraph is that it is also required in flow cost calculation but we can still get rid of the multigraph parameter by self.multigraph = G.is_multigraph() in _DataEssentialsAndFunctions constructor and using DEAF.multigraph instead of multigraph in network_simplex.

Yeah, I thought it'd be nice to get rid of this extra parameter since it is always available via the `G.is_multigraph` method, but I think it's fine for now - no sense in reorganizing the entire code structure for such a minor consideration :).

rossbar reviewed (commented):
This is a very nice PR and I think it offers some significant improvements, namely:
 - Adds testing for `network_simplex` :tada: :rocket: and
 - Refactors the algorithm to use more descriptive variable/attribute names, which makes the entire thing much more readable.

It's also a big PR so will likely take several passes to review :)

This pass is mostly a combination of high-level suggestions or questions, and a few minor nits. I didn't focus too heavily on the implementation of the methods, since they appear to be identical to internally-defined functions with the same name from the previous implementation. I'm curious to hear what you think about some of the questions I have, and please don't hesitate to ask questions of your own if what I said doesn't make sense!

rossbar commented:
I also took the liberty of fixing up a few variable names - the github suggestion feature can get a bit cumbersome for these types of changes.

harshal-dupare commented:
@rossbar I have added it back the `test_google_or_tools_example` with `flowCost=183`. But after these changes, some of the tests are failing and I tried finding why, but no luck so far (I don't think it is related to these recent changes).

rossbar commented:
Thanks for adding back the test @harshal-dupare :)

The CI test errors are unrelated to your changes and should now be fixable by rebasing on `main`.

rossbar commented:
I went ahead and rebased this on main to get the CI green again. Unfortunately this required a force-push, which will make pulling more difficult. @harshal-dupare the easiest way to update your branch locally (assuming your remote is named `origin`) is:

```
git fetch origin
git reset --hard origin/partialfix-for-issue-4660
```

Note that if you have unstaged/uncommitted changes, you should stash first to ensure the hard reset doesn't overwrite them, i.e.:

```
git fetch origin
git stash
git reset --hard origin/partialfix-for-issue-4660
git stash pop
```

If you have any questions or run into unexpected problems, please don't hesitate to ping me!

rossbar commented:
Of course, now their are a whole host of **new** unrelated CI failures. The circleci seems like a CI config issue and the cancelled macOS jobs seem to be on GH's side. Note that none of the new failures are related to the changes in this PR.

harshal-dupare commented:
> I went ahead and rebased this on main to get the CI green again. Unfortunately this required a force-push, which will make pulling more difficult. @harshal-dupare the easiest way to update your branch locally (assuming your remote is named `origin`) is:
> 
> ```
> git fetch origin
> git reset --hard origin/partialfix-for-issue-4660
> ```
> 
> Note that if you have unstaged/uncommitted changes, you should stash first to ensure the hard reset doesn't overwrite them, i.e.:
> 
> ```
> git fetch origin
> git stash
> git reset --hard origin/partialfix-for-issue-4660
> git stash pop
> ```
> 
> If you have any questions or run into unexpected problems, please don't hesitate to ping me!

Thanks @rossbar, I have updated my branch locally.

rossbar commented:
... and I just rebased again to get rid of the *new* unrelated CI failures :joy: . The update procedure will be the same as before - I just want to get the CI back to all-green so we can get this really nice patch in!

dschult reviewed (approved):
I found one typo in a comment :)   
I approve this PR.
